src/pykeen/nn/node_piece.py

```diff
@@ -0,0 +1,683 @@
+            (
+                numpy.ones_like(edge_index[0], dtype=bool),
+                tuple(edge_index),
+            ),
+            shape=(num_entities, num_entities),
+            dtype=bool,
+        )
+        # symmetric + self-loops
+        adjacency = adjacency + adjacency.transpose() + scipy.sparse.eye(num_entities, dtype=bool, format="coo")
+        adjacency = adjacency.tocsr()
+        logger.debug(
+            f"Created sparse adjacency matrix of shape {adjacency.shape} where "
+            f"{format_relative_comparison(part=adjacency.nnz, total=numpy.prod(adjacency.shape))} "
+            f"are non-zero entries.",
+        )
+        return adjacency
+
+    @staticmethod
+    def bfs(
+        anchors: numpy.ndarray,
+        adjacency: scipy.sparse.spmatrix,
+        max_iter: int,
+        k: int,
+    ) -> numpy.ndarray:
+        """
+        Determine the candidate pool using breadth-first search.
+
+        :param anchors: shape: (a,)
+            the anchor node IDs
+        :param adjacency: shape: (n, n)
+            the adjacency matrix
+        :param max_iter:
+            the maximum number of hops to consider
+        :param k:
+            the minimum number of anchor nodes to reach
+
+        :return: shape: (n, a)
+            a boolean array indicating whether anchor $j$ is in the set of $k$ closest anchors for node $i$
+        """
+        num_entities = adjacency.shape[0]
+        # for each entity, determine anchor pool by BFS
+        num_anchors = len(anchors)
+
+        # an array storing whether node i is reachable by anchor j
+        reachable = numpy.zeros(shape=(num_entities, num_anchors), dtype=bool)
+        reachable[anchors] = numpy.eye(num_anchors, dtype=bool)
+
+        # an array indicating whether a node is closed, i.e., has found at least $k$ anchors
+        final = numpy.zeros(shape=(num_entities,), dtype=bool)
+
+        # the output
+        pool = numpy.zeros(shape=(num_entities, num_anchors), dtype=bool)
+
+        old_reachable = reachable
+        for i in range(max_iter):
+            # propagate one hop
+            reachable = adjacency.dot(reachable)
+            # convergence check
+            if (reachable == old_reachable).all():
+                logger.warning(f"Search converged after iteration {i} without all nodes being reachable.")
+                break
+            old_reachable = reachable
+            # copy pool if we have seen enough anchors and have not yet stopped
+            num_reachable = reachable.sum(axis=1)
+            enough = num_reachable >= k
+            mask = enough & ~final
+            logger.debug(
+                f"Iteration {i}: {format_relative_comparison(enough.sum(), total=num_entities)} closed nodes.",
+            )
+            pool[mask] = reachable[mask]
+            # stop once we have enough
+            final |= enough
+            if final.all():
+                break
+        return pool
```

Comment:
@migalkin https://github.com/pykeen/pykeen/pull/755#issuecomment-1023533310

The `pool` *should* contain the `q`-hop neighborhood for the smallest `q` where there are at least `k` anchors inside the neighborhood.

If the implementation is correct, this should be guaranteed by copying the `reachable` array for a given entity

```python
pool[mask] = reachable[mask]
```
for the first time (by `~final`) when there are at least `k` neighbors (by `num_reachable >= k`)
```python
enough = num_reachable >= k
mask = enough & ~final
```
after we copied the `reachable` array for an entity, the entity is closed
```python
final |= enough
```
and thus the `pool` not updated afterwards anymore.

Since the logic is somewhat involved, I'll add an extra test case to verify that :slightly_smiling_face:

Reply:
37e9a14 - seems to work :rocket:

Reply:
Yes yes, I traced the procedure in the debugger and it returns the plausible output 👍 

Let's just keep in mind for the future in case any reproducibility problems occur that in this version anchors within the smallest `q`-hop neighborhood are randomly permuted and not ordered according to the distance

Reply:
Ah, now I get the point - if there are <q hop neighbors, those should always be chosen, and then only the remaining ones be sampled from the =q-hop, right?

Reply:
Yes, correct. 

Suppose we have to select 20 anchors per node, and we found 3 anchors in 1-hop distance, 8 in 2-hop, and 37 in 3-hop. So in order to obtain 20 anchors, we'd take all 1-hop anchors + all 2-hop anchors (11) and sample 9 from 37 of 2-hop. 

That said, I do not know if there would be any performance difference (but surely larger computational overhead) compared to the current mechanism. My gut feeling says no